numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1120,6 +1120,30 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
         goto fail;
     }
 
+    if (!PyArray_CanCastArrayTo(indices, dtype, NPY_SAFE_CASTING)) {
+        if (PyArray_SIZE(indices) == 0) {
+            /*
+             * Unfortunately, if the input array is empty, then indices would
+             * likely have a default dtype of float64, which isn't compatibile.
+             * Force a recast of an empty array.
+             */
+            PyArrayObject *indices_tmp = NULL;
+            Py_INCREF(dtype);
+            indices_tmp = (PyArrayObject*)PyArray_FromArray(indices, dtype,
+                                                            NPY_ARRAY_FORCECAST);
+            if (indices_tmp == NULL) {
+                goto fail;
+            }
+            Py_DECREF(indices);
+            indices = indices_tmp;
+        }
+        else {
+            PyErr_SetString(PyExc_ValueError,
```

Comment:
Admittedly, I am not loving my error message either. I just think that this should be part and parcel with input validation, and so an error message tailored to the function would help track down the source of the problem.

But you are right, it should be a TypeError, not a ValueError.